### PR TITLE
First round of Coulomb contact robustness improvements

### PR DESCRIPTION
**Pull request by bwspenc**
ref #6524 

Note that I have a second commit on deck that adds code to damp the search direction to limit slip reversals, which is really important for this to work in practice.  There are more tests of this capability in that second commit that use this new algorithm, but depend on that damper. I'm not committing it all in a single commit because there are some architectural issues to work out in the second commit. This code is more straightforward.

### CODE REVIEW

File: modules/combined/tests/frictional_contact/sliding_elastic_blocks_2d/sliding_elastic_blocks_2d.i [hunk 1]
```diff
@@ -177,19 +211,19 @@
 
 
 
-  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart'
-  petsc_options_value = 'hypre    boomeramg      101'
+  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
+  petsc_options_value = 'lu     superlu_dist'
```
Comment (jwpeterson) [inline modules/combined/tests/frictional_contact/sliding_elastic_blocks_2d/sliding_elastic_blocks_2d.i]: Please add "superlu = true" in the tests file for this test (and any others that you switch to using `-pc_factor_mat_solver_package superlu_dist`.

Comment (bwspenc) [inline modules/combined/tests/frictional_contact/sliding_elastic_blocks_2d/sliding_elastic_blocks_2d.i]: OK, just did that.

**Comment by permcody** [comment]
@bwspenc - It looks like several of the heavy contact tests failed last night. I reran them just to be sure. I suspect it was this merge. Can you take a look?

https://www.moosebuild.org/job/14253/

**Comment by bwspenc** [comment]
Yes, you're probably right.  I'll look into them.  It looks like one of them is a tolerance issue. @acasagran, can you help look at these as well?

**Pull request merged**